### DISCUSSION TITLE
Multimodels using partial or full data in optimization

### DISCUSSION THREAD

**Discussion by An-xu** [meta-pytorch/botorch#1812] [Q&A] [answered]
Hi, 
I would like to ask about an error I encounter during using botorch. 
Below here is my code. 
def initialize_model(x, y):
    models = []
    for i in range(y.shape[-1]):
        train_objective = y[:, i].unsqueeze(-1)
        if i == 0:
            input_data = x[:, 0:2]
            models.append(
                SingleTaskGP(input_data, train_objective)
            )
        else:
            models.append(
                SingleTaskGP(x, train_objective)
            )
    model = ModelListGP(*models)
    mll = SumMarginalLogLikelihood(model.likelihood, model)
    return mll, model

#%%
def generate_next_candidate(x, y, n_candidates=1):
    mll, model = initialize_model(x, y)
    fit_gpytorch_model(mll)

    pred_list = []
    for i, sub_model in enumerate(model.models):
        input_data = x[:, :2] if i == 0 else x
        with torch.no_grad():
            pred = sub_model.posterior(input_data).mean
        pred_list.append(pred)
    pred = torch.cat(pred_list, dim=1)

    acq_fun_list = []
    for _ in range(n_candidates):
        partitioning = FastNondominatedPartitioning(ref_point=ref_point, Y=pred)
        acq_fun = ExpectedHypervolumeImprovement(
            model=model, ref_point=ref_point, partitioning=partitioning
        )
        acq_fun_list.append(acq_fun)

    # modify bounds based on input dimensionality
    bounds_list = []
    for i in range(len(model.models)):
        d = 2 if i == 0 else 4
        bounds = torch.zeros(2, d)
        bounds[1] = 1.0
        bounds_list.append(bounds)

    candidates_list = []
    for i, acq_fun in enumerate(acq_fun_list):
        candidates, _ = optimize_acqf(
            acq_function=acq_fun,
            bounds=bounds_list[i],
            q=n_candidates,
            num_restarts=1000,
            batch_initial_conditions=None,
            raw_samples=11000,
        )
        candidates_list.append(candidates)

    return candidates_list, pred, model
    
x = torch.tensor(data[['ratio1', 'ratio2','ratio3','ratio4']].values, dtype=torch.float)
y = torch.tensor(data[['PF', 'Printing']].values, dtype=torch.float)

for i in range(n_iter):
    print(f"Iteration {i}")
    candidates, pred, model = generate_next_candidate(train_x_norm, train_y_norm, n_candidates=1)
    
    
    
When I iterate the function, I have an error :RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 4 but got size 2 for tensor number 1 in the list.
What approach I should take to fix the error? 

Thanks a lot for your help!!

**Comment by An-xu** [comment]
Hi, Since the model need to remain the same input dimension, I set the non-related variables to 0. The function is written below. Let me know if there are better solutions. 
Thank you so much for your help! 

def initialize_model(train_x, train_y):
    models = []
    for i in range(train_y.shape[1]):
        train_objective = train_y[:, i].unsqueeze(-1)
        if i == 0:
            input_data = torch.cat((train_x[:,0:2],torch.zeros(train_x.shape[0],2)),1)
            models.append(
                SingleTaskGP(input_data, train_objective)
            )
        else:
            models.append(
                SingleTaskGP(train_x, train_objective)
            )
    model = ModelListGP(*models)
    mll = SumMarginalLogLikelihood(model.likelihood, model)
    return mll, model

**Reply by Balandat** [reply, marked as answer]
So one option you have is to use the [`FilterFeatures`](https://github.com/pytorch/botorch/blob/main/botorch/models/transforms/input.py#L1216) input transform that just picks out the desired features for the model. That way you don't have to introduce irrelevant zero features for one of your models.

**Reply by An-xu** [reply]
Thank you! That is really helpful!

**Comment by An-xu** [comment]
Hi Balandat, 
Thanks again for your reply! 
After filtering the feature, I encounter a problem during the optimization process, I couldn't set the bound separately. It has the error: 
RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 4 but got size 3 for tensor number 1 in the list.

def initialize_model(x, y, var=None):
    models = []
    for i in range(y.shape[1]):
        if  i == 2:
            train_x = x
        else:
            feature_indices = torch.tensor([0, 1, 2])  # indices of the features we want to keep
            transform = FilterFeatures(feature_indices)
            train_x = transform(x)
        train_objective = y[:, i].unsqueeze(-1)
        models.append(
            HeteroskedasticSingleTaskGP(train_x, train_objective, var[:, i].unsqueeze(-1))
            )

    model = ModelListGP(*models)
    mll = SumMarginalLogLikelihood(model.likelihood, model)
    return mll, model
    
def optimize_qehvi_and_get_observation(model, x, y, ref_point, n_candidates, sampler):
    with torch.no_grad():
            feature_indices = torch.tensor([0, 1, 2])  # indices of the features we want to keep
            transform = FilterFeatures(feature_indices)
            x_val_transformed = transform(x)
            preds_0 = model.models[0].posterior(x_val_transformed)
            preds_1 = model.models[1].posterior(x_val_transformed)
            preds_2 = model.models[2].posterior(x)
            preds = torch.stack((preds_0.mean, preds_1.mean, preds_2.mean), dim=1).squeeze(-1)
    print("preds shape:", preds.shape)
    acq_fun_list = []

    partitioning = FastNondominatedPartitioning(ref_point=ref_point,Y=preds)

    acq_fun =qExpectedHypervolumeImprovement(model=model, ref_point=ref_point, partitioning=partitioning, sampler=sampler)
    # acq_fun =ExpectedHypervolumeImprovement(model=model, ref_point=ref_point, partitioning=partitioning)

    bounds = torch.zeros(2, 3)
    bounds[1] = 1
    candidates, _ = optimize_acqf(
        acq_function=acq_fun,
        bounds= bounds,
        q=n_candidates,
        num_restarts=100,
        batch_initial_conditions=None,
        raw_samples=1100,
    )
    return candidates, preds
